app/src/main/java/org/mozilla/focus/telemetry/AppLaunchMethod.java

```diff
@@ -0,0 +1,78 @@
+package org.mozilla.focus.telemetry;
+
+import android.content.Intent;
+import android.support.annotation.NonNull;
+
+import org.mozilla.focus.utils.SafeIntent;
+
+public enum AppLaunchMethod {
+
+    LAUNCHER() {
+        @Override
+        boolean match(@NonNull SafeIntent intent) {
+            return Intent.ACTION_MAIN.equals(intent.getAction());
+        }
+
+        @Override
+        public void sendLaunchTelemetry() {
+            TelemetryWrapper.launchByAppLauncherEvent();
+        }
+    },
+
+    HOME_SCREEN_SHORTCUT() {
+        @Override
+        boolean match(@NonNull SafeIntent intent) {
+            return Intent.ACTION_VIEW.equals(intent.getAction())
+                    && intent.getBooleanExtra(EXTRA_HOME_SCREEN_SHORTCUT, false);
+        }
+
+        @Override
+        public void sendLaunchTelemetry() {
+            TelemetryWrapper.launchByHomeScreenShortcutEvent();
+        }
+    },
+
+    TEXT_SELECTION_SEARCH() {
+        @Override
+        boolean match(@NonNull SafeIntent intent) {
+            return Intent.ACTION_VIEW.equals(intent.getAction())
+                    && intent.getBooleanExtra(EXTRA_TEXT_SELECTION, false);
+        }
+
+        @Override
+        public void sendLaunchTelemetry() {
+            TelemetryWrapper.launchByTextSelectionSearchEvent();
+        }
+    },
+
+    EXTERNAL_APP() {
+        @Override
+        boolean match(@NonNull SafeIntent intent) {
+            return Intent.ACTION_VIEW.equals(intent.getAction());
+        }
+
+        @Override
+        public void sendLaunchTelemetry() {
+            TelemetryWrapper.launchByExternalAppEvent();
+        }
+    };
+
+    public static final String EXTRA_TEXT_SELECTION = "text_selection";
+    public static final String EXTRA_HOME_SCREEN_SHORTCUT = "shortcut";
+
+    abstract boolean match(@NonNull SafeIntent intent);
+
+    public abstract void sendLaunchTelemetry();
+
+    public static AppLaunchMethod parse(SafeIntent intent) {
+        if (intent != null) {
+            for (AppLaunchMethod method : AppLaunchMethod.values()) {
```

Comment:
With this implementation, we'll need to make sure broader match() appears later in the list of different AppLaunchMethod types. Maybe stating this in `abstract boolean match(@NonNull SafeIntent intent)` or `public enum AppLaunchMethod {`?